torchao/prototype/quantization/safetensor.py

```diff
@@ -0,0 +1,206 @@
+        - metadata: Dictionary with metadata needed to reconstruct the tensor
+        - tensors_dict: Dictionary with tensors to save
+    """
+    metadata = {}
+    tensors_dict = {}
+
+    if tensor.__class__.__name__ == "Float8Tensor":
+        metadata["tensor_type"] = "Float8Tensor"
+        for item, value in tensor.__dict__.items():
+            if isinstance(value, granularity.PerRow):
+                metadata["granularity"] = "PerRow"
+            elif isinstance(value, QuantizeTensorToFloat8Kwargs):
+                metadata["dtype"] = "torch.float8_e4m3fn"
+            elif item == "qdata":
+                tensors_dict["qdata"] = value
+                metadata["qdata_device"] = (
+                    "cuda:0" if value.device == torch.device("cuda:0") else "cpu"
+                )
+            elif item == "scale":
+                tensors_dict["scale"] = value
+                metadata["scale_device"] = (
+                    "cuda:0" if value.device == torch.device("cuda:0") else "cpu"
+                )
+            elif value:
+                metadata[item] = json.dumps(value)
+
+    return metadata, tensors_dict
+
+
+def save_tensor_subclass_dict(
+    tensor_dict: Dict[str, Dict[str, torch.Tensor]],
+    file_path: str,
+    additional_metadata: Optional[Dict[str, str]] = None,
+):
+    """
+    Save a dictionary of tensor subclasses with appropriate metadata.
+
+    Args:
+        tensor_dict: Dictionary of tensor subclasses to save, with keys as tensor names
+        file_path: Path where to save the tensors
+        additional_metadata: Optional additional metadata to include
+    """
+
+    combined_metadata = {}
+    combined_tensors_dict = {}
+
+    for tensor_name, tensor in tensor_dict.items():
+        if tensor.__class__.__name__ == "Tensor":
+            tensors_dict = {"data": tensor}
+            metadata = {"tensor_type": "Tensor"}
+            metadata["device"] = (
+                "cuda:0" if tensor.device == torch.device("cuda:0") else "cpu"
+            )
+        else:
+            metadata, tensors_dict = create_metadata_for_tensor_subclass(tensor)
+
+        # Clone tensors to avoid memory sharing issues
+        prefixed_tensors_dict = {
+            f"{tensor_name}:{key}": (
+                value.detach().clone() if isinstance(value, torch.Tensor) else value
+            )
+            for key, value in tensors_dict.items()
+        }
+
+        for key, value in metadata.items():
+            combined_metadata[f"{tensor_name}:{key}"] = value
+
+        combined_tensors_dict.update(prefixed_tensors_dict)
+
+    combined_metadata["tensor_names"] = json.dumps(list(tensor_dict.keys()))
+
+    if additional_metadata:
+        combined_metadata.update(additional_metadata)
+
+    save_file(combined_tensors_dict, file_path, metadata=combined_metadata)
+    print(f"Saved {len(tensor_dict)} tensor subclasses to {file_path} with metadata")
+
+
+if __name__ == "__main__":
```

Comment:
this should be in a separate test file in https://github.com/pytorch/ao/tree/main/test/prototype

eg. test/prototype/safetensors/test_safetensors.py